### PR TITLE
File System API

### LINKED ISSUE #329: Add simple filesystem backend
**- Do you want to request a *feature* or report a *bug*?**

New feature.

**- What is the current behavior?**

Currently there is no backend which uses the filesystem. This would be useful for local work and testing.

**- What is the expected behavior?**

There should be a backend available which calls a simple API that allows reading and writing from the filesystem. Netlify CMS should allow you to spin up a local filesystem-backed instance via an `npm` task.

### PR DESCRIPTION
**Pull request by talves**
**- Summary**

Create an end point file system api middleware for the express server (webpack-dev-server) to be able to create a backend api in the CMS to update the example site locally.

1. Checkout the branch `yarn` and `yarn start`
2. The devServer will list two paths. The root path to the project location and the path to the site directory. These paths should be correct.

Example on windows:
```
Root path is D:\Git\Netlify\netlify-cms-origin\
Site path is D:\Git\Netlify\netlify-cms-origin\example
```

Closes #329 

**- Test plan**

- [ ] Needs a test on POSIX systems
- [ ] Needs a test on MAC
- [ ] Windows has been tested but another look would be good

**- Description for the changelog**
Features:

- Able to run against the `/api` path at two endpoints `/api/file/:id` and `/api/files/:path`
- The api will later be able to run as a stand alone middleware plugin for a local express server with no changes (written)

### CODE REVIEW

**Comment by talves** [comment]
@Benaiah  @erquhart I will explain as much as I can on this PR, so you can test quicker.

I have created a new example structure at the root of `example` the way Shawn likes it 😄 It has the same paths and folders as the orphan branch `cms-test-backend`.

## Structure for deploy:
I created a `config.yml` file in the assets folder to use when a deploy preview kicks off from now on that will point to the orphan branch. All someone has to do on a test branch that needs a different layout is put their copy of the config in there and it will now get copied over to the root `example/config.yml` when the PR deploy preview is created.

## API
The new ***backend*** API is called `file-system` in the cms. It targets the endpoints in the API middleware in the `scripts/fs` folder. When a webpack dev server is started, the api routes are added to the server.

## Notes:
I think all the backend API's should be plugins at some point for the cms, so we are not carrying around unnecessary code.

The failures on read or write do make the cms fail elegantly. Any errors will show up in the inspect of the page as 500 errors.

Hope this all makes sense and code like you are on :fire:

File: example/assets/config.yml [hunk 1]
```diff
@@ -0,0 +1,166 @@
+backend:
+  name: github
+  repo: netlify/netlify-cms
+  branch: cms-test-backend
```
Comment (talves) [inline example/assets/config.yml]: This file is used for copy on deploy preview!
The `cms-test-backend` branch is an orphan branch so new changes can be tested on deploy preview.

File: example/config.yml [hunk 2]
```diff
@@ -1,5 +1,5 @@
 backend:
-  name: test-repo
+  name: file-system
```
Comment (talves) [inline example/config.yml]: The default config file will allow for devServer (webpack) to start testing immediately on local branch checkout. The test-repo backend is not really needed once this solution is in place.

**Comment by tech4him1** [comment]
> I think all the backend API's should be plugins at some point for the cms, so we are not carrying around unnecessary code.

Interesting, I actually suggested that a while ago: https://github.com/netlify/netlify-cms/issues/465#issuecomment-313405143

File: example/assets/preview/config.yml [hunk 3]
```diff
@@ -0,0 +1,166 @@
+backend:
+  name: github
+  repo: netlify/netlify-cms
```
Comment (tech4him1) [inline example/assets/preview/config.yml]: If we use the CMS repo on the GitHub backend for deploy previews, contributors that do not have write permissions will not be able to use our deploy previews. Right?

File: webpack.dev.js [hunk 4]
```diff
@@ -60,5 +61,6 @@ module.exports = merge.smart(require('./webpack.base.js'), {
     historyApiFallback: true,
     disableHostCheck: true,
     headers: {"Access-Control-Allow-Origin": "*"},
+    setup: fileSystemAPIPlugin,
```
Comment (tech4him1) [inline webpack.dev.js]: > The `setup` option is deprecated and will be removed in v3. Please update your config to use `before`

**Comment by talves** [comment]
I will be closing this and replacing with PR #786 

I will address comments there from now on.

**Pull request closed without merging**